### PR TITLE
Implement `DATE` type in the backend

### LINKED ISSUE #425: Handle `DATE` data type in the backend
This issue is to ensure that Mathesar can handle the Postgres `DATE` data type.

As part of this issue, we need to ensure that:
- Users can use the API to change a column to `DATE` if it's possible to do so.
- Automatic type inference during file import suggests `DATE` when it makes sense to do so. 
  - We could use something like https://dateparser.readthedocs.io/

**Additional Context** 
- #250 (parent issue)
- See PR #368 for code that calculates which types are allowed for a column.
- This issue was previously specced out in #128

### PR DESCRIPTION
**Pull request by eito-fis**
Fixes #425

Implements the `DATE` type for use in column alterations in the backend.

**Technical details**
Uses the default postgres string -> date conversion, see [here](https://www.postgresql.org/docs/13/datatype-datetime.html#DATATYPE-DATETIME-INPUT).

**Screenshots**

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/tests/types/test_alteration.py [hunk 1]
```diff
@@ -327,6 +336,19 @@
...
             },
+            DATE: {
+                VALID: [
+                    ("1999-01-18", date(1999, 1, 18)),
+                    ("1/18/1999", date(1999, 1, 18)),
+                    ("jan-1999-18", date(1999, 1, 18)),
+                    ("19990118", date(1999, 1, 18)),
+                ],
+                INVALID: [
+                    "18/1/1999",
```
Comment (kgodey) [inline db/tests/types/test_alteration.py:347]: This is a valid date in a non-US format (this is how I grew up writing dates). Does Postgres not parse this?

**Review by kgodey — COMMENTED** [review]
@eito-fis This looks pretty good. I added a question.

Also, I think it would be good to add some tests for inferring dates on import.

**Review by powellc — APPROVED** [review]
Nice work, @eito-fis ! I too am curious about the supported date formats and wonder if there are any others that Postgres doesn't handle 🤔  But otherwise I think it's good to merge 🚀

File: db/tests/types/test_alteration.py (hunk 1 above)
Comment (eito-fis) [reply db/tests/types/test_alteration.py:347]: This is dependent on the server [`DateStyle`](https://www.postgresql.org/docs/13/runtime-config-client.html#GUC-DATESTYLE) option, which defaults to MDY. As its a server wide setting, as of now I don't think we can support parsing multiple different formats. If it is something we want, we might be able to do it with a custom type that has a `format` type option? I'm dealing with a similar problem for the `TIME` type right now, so I'll know how hard it'll be once that's done.

Comment (kgodey) [reply db/tests/types/test_alteration.py:347]: Okay, let's not worry about it now, we can open a new issue for improving parsing later if needed.

I'll review again once there's a test for ensuring that dates are inferred correctly on import. Thanks!

Comment (kgodey) [reply db/tests/types/test_alteration.py:347]: Also please add a comment in the code about the `DateStyle` option.

**Comment by eito-fis** [comment]
Updated the inference tests with an additional case where we convert `STRING` -> `DATE`.

**Review by kgodey — APPROVED** [review]

**Comment by kgodey** [comment]
Looks good, thanks! I think we should add a test at some point at the API layer of importing a CSV with all the data types that we infer and make sure they are inferred correctly. Perhaps you could add this in a future PR?

**Pull request merged**